qualtran/bloqs/utils.py

```diff
@@ -0,0 +1,38 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import Union
+
+import numpy as np
+import sympy
+from cirq._doc import document
+
+SymbolicFloat = Union[float, sympy.Expr]
+document(SymbolicFloat, """A floating point value or a sympy expression.""")
+
+SymbolicInt = Union[int, sympy.Expr]
+document(SymbolicFloat, """A floating point value or a sympy expression.""")
+
+
+def log2(x: SymbolicFloat) -> SymbolicFloat:
+    from sympy.codegen.cfunctions import log2
+
+    if not isinstance(x, sympy.Basic):
+        return np.log2(x)
+    return log2(x)
+
+
+def ceil(x: SymbolicFloat) -> SymbolicInt:
+    if not isinstance(x, sympy.Basic):
+        return int(np.ceil(x))
+    return sympy.ceiling(x)
```

Comment:
why not always use the sympy functions? they should work on concrete values

Reply:
It's because they return sympy types and it's hard to distinguish whether a bloq is parameterized or not (eg: using cirq.is_parameterized(val) which uses a bunch of isinstance checks)

```python
In [7]: isinstance(sympy.log(2), numbers.Number)
Out[7]: False

In [8]: type(sympy.log(2))
Out[8]: log
```

Reply:
Just for completeness: sympy has evaluation rules which keeps `sympy.log(2)` as symbolic because there isn't a way of representing the result with the same precision as the input. If you use floats or in the `ceiling` example: it gets evaluated.

```python
isinstance(sympy.log(2.), numbers.Number)
Out[15]: True
isinstance(sympy.ceiling(2), numbers.Number)
Out[16]: True
isinstance(sympy.ceiling(2.), numbers.Number)
Out[17]: True
```

https://docs.sympy.org/latest/guides/custom-functions.html

It might be less surprising for people if we say "for symbolic-supporting gates, the symbolic rules of sympy are used" (note: this is also how mathematica does it)

Reply:
I bet there will be cases where we want things to always evaluate numerically if possible but I bet there will be cases where someone will want to keep something as a sympy expression and not have it considered "parameterized". The "if it's sympy then it's a parameterized gate" logic is not quite right. 

For the first case: instead of these functions with "if" it might be better to keep things sympy until we need them to be numbers and then do the sympy evaluate-as-float button. In mathematica it's `// N` I'm sure such a thing exists in sympy too.

Reply:
update: it's `N` in sympy too https://docs.sympy.org/latest/modules/evalf.html